### PR TITLE
benchmark_signature_verification

### PR DESCRIPTION
**Pull request by ilincaseby**
## Reasoning behind the pull request
- Measure performance regarding the approach of implementing the signature verifier method.
- Discussing the future implementation that'll be put in production.
- Implementing parallel signature verification(with throttler)
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: consensus/spos/bls/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	for i := 0; i < b.N; i++ {
+		b.StartTimer()
+		invalidSigners, err := sr.VerifyNodesOnAggSigFailAux()
+		b.StopTimer()
+		require.Nil(b, err)
+		require.NotNil(b, invalidSigners)
+	}
+}
+
+func BenchmarkSubroundEndRound_VerifyNodesOnAggSigFailTimeParallelThrottle(b *testing.B) {
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/benchmark_test.go:172]: we would end up only with this benchmark, probably the others need to be removed.

File: consensus/spos/bls/errors.go [hunk 2]
```diff
@@ -4,3 +4,13 @@ import "errors"
 
 // ErrNilSentSignatureTracker defines the error for setting a nil SentSignatureTracker
 var ErrNilSentSignatureTracker = errors.New("nil sent signature tracker")
+
+var JobIsNotDoneError error = errors.New("job is not done")
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/errors.go:8]: Please follow the model of errors names from the file
e.g all error should start with `Err`

also need comment on exported errors

File: consensus/spos/bls/export_test.go [hunk 3]
```diff
@@ -366,3 +368,15 @@ func (sr *subroundStartRound) ChangeEpoch(epoch uint32) {
 func (sr *subroundStartRound) IndexRoundIfNeeded(pubKeys []string) {
 	sr.indexRoundIfNeeded(pubKeys)
 }
+
+func (sr *subroundEndRound) SignatureVerification(wg *sync.WaitGroup, i int, pk string, invalidPubKey *[]string, mutex *sync.Mutex, sigShare []byte, mutexBool *sync.Mutex, errorReturned *error) {
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/export_test.go:372]: comment on exported methods.

File: consensus/spos/bls/subroundEndRound.go [hunk 4]
```diff
@@ -21,13 +22,19 @@ import (
 
 const timeBetweenSignaturesChecks = time.Millisecond * 5
 
+type ErrSigVerificationPair struct {
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:25]: I don't think we need this, can be removed.

File: consensus/spos/bls/subroundEndRound.go [hunk 5]
```diff
@@ -553,6 +565,119 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+		case <-time.After(time.Millisecond):
+			continue
+		case <-ctx.Done():
+			return spos.ErrTimeIsOut
+		}
+	}
+	return nil
+}
+
+func (sr *subroundEndRound) signatureVerification(wg *sync.WaitGroup, i int, pk string, invalidPubKey *[]string, mutex *sync.Mutex, sigShare []byte, mutexBool *sync.Mutex, errorReturned *error) {
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:583]: please refactor, removing the wg, invalidPubKey, mutex,  mutexBool, errorReturned from the arguments list.
you can do the rest in an anonymous function that is called on a goroutine.

File: consensus/spos/bls/subroundEndRound.go [hunk 6]
```diff
@@ -553,6 +565,119 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+		go sr.signatureVerification(wg, i, pk, &invalidPubKeys, mutex, sigShare, mutexBool, &errorReturned)
+	}
+	wg.Wait()
+	if errorReturned != nil {
+		return nil, errorReturned
+	}
+	return invalidPubKeys, nil
+}
+
+func (sr *subroundEndRound) verifyNodesOnAggSigFailAuxThrottle(ctx context.Context) ([]string, error) {
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:642]: we should keep only the version with throttle, but give it the name `verifyNodesOnAggSigFail`

File: consensus/spos/bls/subroundEndRound_test.go [hunk 7]
```diff
@@ -13,21 +13,23 @@ import (
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
 	"github.com/multiversx/mx-chain-go/common"
 	"github.com/multiversx/mx-chain-go/consensus"
 	"github.com/multiversx/mx-chain-go/consensus/mock"
 	"github.com/multiversx/mx-chain-go/consensus/spos"
 	"github.com/multiversx/mx-chain-go/consensus/spos/bls"
 	"github.com/multiversx/mx-chain-go/dataRetriever/blockchain"
+	mock2 "github.com/multiversx/mx-chain-go/dataRetriever/mock"
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound_test.go:25]: rename the alias to dataRetrieverMocks

File: consensus/spos/sposFactory/sposFactory_test.go [hunk 8]
```diff
@@ -5,17 +5,19 @@ import (
...
 	"github.com/multiversx/mx-chain-core-go/core"
 	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/stretchr/testify/assert"
+
 	"github.com/multiversx/mx-chain-go/config"
 	"github.com/multiversx/mx-chain-go/consensus"
 	"github.com/multiversx/mx-chain-go/consensus/mock"
 	"github.com/multiversx/mx-chain-go/consensus/spos"
 	"github.com/multiversx/mx-chain-go/consensus/spos/sposFactory"
+	mock2 "github.com/multiversx/mx-chain-go/dataRetriever/mock"
```
Comment (AdoAdoAdo) [inline consensus/spos/sposFactory/sposFactory_test.go:15]: you can rename the alias to dataRetrieverMocks

File: factory/consensus/consensusComponents.go [hunk 9]
```diff
@@ -30,14 +33,14 @@ import (
...
 	"github.com/multiversx/mx-chain-go/update"
-	logger "github.com/multiversx/mx-chain-logger-go"
-	"github.com/multiversx/mx-chain-storage-go/timecache"
 )
 
 var log = logger.GetOrCreate("factory")
 
 const defaultSpan = 300 * time.Second
 
+const numSignatureGoRoutinesThrottler = 64
```
Comment (AdoAdoAdo) [inline factory/consensus/consensusComponents.go:42]: this I think can be removed, as you would have access to the same constant used for signing goroutines throttler

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	sort.Strings(keys)
+
+	return keys
+}
+
+// Benchmark on measuring performance
+func BenchmarkSubroundEndRound_VerifyNodesOnAggSigFailTime(b *testing.B) {
+	b.ResetTimer()
+	b.StopTimer()
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:45]: add an empty line after

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	ctx, cancel := context.WithCancel(context.TODO())
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:46]: to be easier to read + less prone to errors, after this line add a :point_down: and remove the cancel call at the end of the test
```go
defer func() {
   cancel()
}
```

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+	suite := mcl.NewSuiteBLS12()
+	kg := signing.NewKeyGenerator(suite)
+	mapKeys := make(map[string]crypto2.PrivateKey)
+
+	for i := uint16(0); i < 400; i++ {
+		sk, pk := kg.GeneratePair()
+
+		pubKey, _ := pk.ToByteArray()
+		mapKeys[string(pubKey)] = sk
+	}
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:64]: can be extracted in a function

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	"context"
+	"sort"
+	"testing"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-crypto-go/signing"
+	"github.com/multiversx/mx-chain-crypto-go/signing/mcl"
+	"github.com/stretchr/testify/require"
+
+	crypto2 "github.com/multiversx/mx-chain-crypto-go"
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:13]: this should stay as crypto simple

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	multisig2 "github.com/multiversx/mx-chain-crypto-go/signing/mcl/multisig"
+	"github.com/multiversx/mx-chain-crypto-go/signing/multisig"
+
+	"github.com/multiversx/mx-chain-go/consensus/spos/bls"
+	dataRetrieverMocks "github.com/multiversx/mx-chain-go/dataRetriever/mock"
+
+	"github.com/multiversx/mx-chain-go/common"
+	"github.com/multiversx/mx-chain-go/factory/crypto"
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:21]: add an alias as factoryCrypto

File: consensus/spos/bls/subroundEndRound.go [hunk 15]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+			continue
+		case <-ctx.Done():
+			return spos.ErrTimeIsOut
+		}
+	}
+	return nil
+}
+
+// Method that implements parallel signature verification
+func (sr *subroundEndRound) signatureVerification(i int, pk string, sigShare []byte) error {
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:582]:
```suggestion
func (sr *subroundEndRound) verifySignature(i int, pk string, sigShare []byte) error {
```

File: consensus/spos/bls/subroundEndRound.go [hunk 16]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
 	return bitmap, sig, nil
 }
 
-func (sr *subroundEndRound) verifyNodesOnAggSigFail() ([]string, error) {
+// Method to let thread continue its execution so that another goroutine can be called
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:563]: usually we start comments with function name, valid for all

File: consensus/spos/bls/subroundEndRound.go [hunk 17]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+func (sr *subroundEndRound) checkGoRoutinesThrottler(ctx context.Context) error {
+	for {
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:566]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 18]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+		case <-ctx.Done():
+			return spos.ErrTimeIsOut
+		}
+	}
+	return nil
+}
+
+// Method that implements parallel signature verification
+func (sr *subroundEndRound) signatureVerification(i int, pk string, sigShare []byte) error {
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:583]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 19]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+	isSuccessful := true
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:585]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 20]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+	err := sr.SigningHandler().VerifySignatureShare(uint16(i), sigShare, sr.GetData(), sr.Header.GetEpoch())
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:587]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 21]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+	if err != nil {
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:589]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 22]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+			return spos.ErrTimeIsOut
+		}
+	}
+	return nil
+}
+
+// Method that implements parallel signature verification
+func (sr *subroundEndRound) signatureVerification(i int, pk string, sigShare []byte) error {
+
+	isSuccessful := true
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:584]: no need for this boolean.. in case of err, return err and add another log.trace, in case of success, simply return nil

File: consensus/spos/bls/subroundEndRound.go [hunk 23]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+	return err
+}
+
+func (sr *subroundEndRound) verifyNodesOnAggSigFail(ctx context.Context) ([]string, error) {
 	invalidPubKeys := make([]string, 0)
 	pubKeys := sr.ConsensusGroup()
+	wg := &sync.WaitGroup{}
+	mutex := &sync.Mutex{}
 
 	if check.IfNil(sr.Header) {
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:615]: this should be moved as first thing on this function

File: consensus/spos/bls/subroundEndRound.go [hunk 24]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
 		return nil, spos.ErrNilHeader
 	}
 
 	for i, pk := range pubKeys {
 		isJobDone, err := sr.JobDone(pk, SrSignature)
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:621]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 25]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
 		if err != nil || !isJobDone {
 			continue
 		}
 
 		sigShare, err := sr.SigningHandler().SignatureShare(uint16(i))
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:627]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 26]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
 		if err != nil {
 			return nil, err
 		}
 
-		isSuccessfull := true
-		err = sr.SigningHandler().VerifySignatureShare(uint16(i), sigShare, sr.GetData(), sr.Header.GetEpoch())
+		err = sr.checkGoRoutinesThrottler(ctx)
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:633]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 27]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
-				pk,
-				spos.GetConsensusTopicID(sr.ShardCoordinator()),
-				decreaseFactor,
-			)
+		wg.Add(1)
 
-			invalidPubKeys = append(invalidPubKeys, pk)
-		}
+		go func(i int, pk string, wg *sync.WaitGroup, sigShare []byte) {
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:643]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 28]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+			defer sr.signatureThrottler.EndProcessing()
+			defer wg.Done()
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:645]:
```suggestion
			defer func () {
				sr.signatureThrottler.EndProcessing()
				wg.Done()
			}
```

File: consensus/spos/bls/subroundEndRound.go [hunk 29]
```diff
@@ -553,48 +560,100 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
 
-		log.Trace("verifyNodesOnAggSigVerificationFail: verifying signature share", "public key", pk, "is successfull", isSuccessfull)
+			errSigVerification := sr.signatureVerification(i, pk, sigShare)
+
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:648]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound.go [hunk 30]
```diff
@@ -621,7 +680,9 @@ func (sr *subroundEndRound) getFullMessagesForInvalidSigners(invalidPubKeys []st
 }
 
 func (sr *subroundEndRound) handleInvalidSignersOnAggSigFail() ([]byte, []byte, error) {
-	invalidPubKeys, err := sr.verifyNodesOnAggSigFail()
+	ctx, cancel := context.WithCancel(context.TODO())
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:683]:
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
```

Comment (sstanculeanu) [reply consensus/spos/bls/subroundEndRound.go:683]: with this implementation, I don't think timeout will ever be reached, perhaps the node can get stuck
reason: 
- cancel is called when verifyNodesOnAggSigFail is finished
- verifyNodesOnAggSigFail has a blocking call to checkGoRoutinesThrottler
- checkGoRoutinesThrottler ends in 2 conditions:
   - the signature can be sent
   - timeout

maybe a better approach here would be to use a context with timeout = with round duration:
```go
ctx, cancel := context.WithTimeout(context.Background(), sr.RoundHandler().TimeDuration())
```

File: factory/consensus/consensusComponents.go [hunk 31]
```diff
@@ -265,12 +265,12 @@ func (ccf *consensusComponentsFactory) Create() (*consensusComponents, error) {
 	if err != nil {
 		return nil, err
 	}
-
 	signatureThrottler, err := throttler.NewNumGoRoutinesThrottler(numSignatureGoRoutinesThrottler)
 	if err != nil {
 		return nil, err
 	}
 
+
```
Comment (sstanculeanu) [inline factory/consensus/consensusComponents.go:273]:
```suggestion
```

File: consensus/spos/bls/subroundEndRound_test.go [hunk 32]
```diff
@@ -361,6 +411,7 @@ func TestSubroundEndRound_NewSubroundEndRoundNilSyncTimerShouldFail(t *testing.T
 		&statusHandler.AppStatusHandlerStub{},
 		&testscommon.SentSignatureTrackerStub{},
 		&mock.SposWorkerMock{},
+		&dataRetrieverMocks.ThrottlerStub{},
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound_test.go:414]: missing test for nil throttler

File: consensus/spos/bls/blsWorker_test.go [hunk 33]
```diff
@@ -86,6 +86,40 @@ func initConsensusStateWithKeysHandlerWithGroupSizeWithRealKeys(keysHandler cons
 	return cns
 }
 
+func initConsensusStateWithArgsVerifySignature(keysHandler consensus.KeysHandler, keys []string) *spos.ConsensusState {
+	numberOfKeys := len(keys)
+	eligibleNodesPubKeys := make(map[string]struct{})
```
Comment (sstanculeanu) [inline consensus/spos/bls/blsWorker_test.go:91]:
```suggestion
	eligibleNodesPubKeys := make(map[string]struct{}, numberOfKeys)
```

File: consensus/spos/bls/blsWorker_test.go [hunk 34]
```diff
@@ -86,6 +86,40 @@ func initConsensusStateWithKeysHandlerWithGroupSizeWithRealKeys(keysHandler cons
...
+	for _, key := range keys {
+		eligibleNodesPubKeys[key] = struct{}{}
+	}
+	indexLeader := 1
```
Comment (sstanculeanu) [inline consensus/spos/bls/blsWorker_test.go:95]:
```suggestion
	}
	
	indexLeader := 1
```

File: consensus/spos/bls/blsWorker_test.go [hunk 35]
```diff
@@ -86,6 +86,40 @@ func initConsensusStateWithKeysHandlerWithGroupSizeWithRealKeys(keysHandler cons
...
+	rcns, _ := spos.NewRoundConsensus(
+		eligibleNodesPubKeys,
+		numberOfKeys,
+		keys[indexLeader],
+		keysHandler,
+	)
+	rcns.SetConsensusGroup(keys)
+	rcns.ResetRoundState()
+	pBFTThreshold := numberOfKeys*2/3 + 1
```
Comment (sstanculeanu) [inline consensus/spos/bls/blsWorker_test.go:104]:
```suggestion
	rcns.ResetRoundState()
	
	pBFTThreshold := numberOfKeys*2/3 + 1
```

File: consensus/spos/bls/blsWorker_test.go [hunk 36]
```diff
@@ -86,6 +86,40 @@ func initConsensusStateWithKeysHandlerWithGroupSizeWithRealKeys(keysHandler cons
...
+	pBFTFallbackThreshold := numberOfKeys*1/2 + 1
+	rthr := spos.NewRoundThreshold()
+	rthr.SetThreshold(1, 1)
+	rthr.SetThreshold(2, pBFTThreshold)
+	rthr.SetFallbackThreshold(1, 1)
+	rthr.SetFallbackThreshold(2, pBFTFallbackThreshold)
+	rstatus := spos.NewRoundStatus()
```
Comment (sstanculeanu) [inline consensus/spos/bls/blsWorker_test.go:111]:
```suggestion
	rthr.SetFallbackThreshold(2, pBFTFallbackThreshold)
	
	rstatus := spos.NewRoundStatus()
```

File: consensus/spos/bls/blsWorker_test.go [hunk 37]
```diff
@@ -86,6 +86,40 @@ func initConsensusStateWithKeysHandlerWithGroupSizeWithRealKeys(keysHandler cons
...
+	rstatus.ResetRoundStatus()
+	cns := spos.NewConsensusState(
+		rcns,
+		rthr,
+		rstatus,
+	)
+	cns.Data = []byte("X")
+	cns.RoundIndex = 0
+	return cns
```
Comment (sstanculeanu) [inline consensus/spos/bls/blsWorker_test.go:120]:
```suggestion
	cns.RoundIndex = 0
	
	return cns
```

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 38]
```diff
@@ -0,0 +1,119 @@
...
+	"sort"
+	"testing"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-crypto-go/signing"
+	"github.com/multiversx/mx-chain-crypto-go/signing/mcl"
+	"github.com/stretchr/testify/require"
+
+	crypto "github.com/multiversx/mx-chain-crypto-go"
+	multisig2 "github.com/multiversx/mx-chain-crypto-go/signing/mcl/multisig"
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:14]:
```suggestion
	mclMultisig "github.com/multiversx/mx-chain-crypto-go/signing/mcl/multisig"
```

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 39]
```diff
@@ -0,0 +1,119 @@
...
+
+	sort.Strings(keys)
+
+	return keys
+}
+
+func generateKeyPairs(kg crypto.KeyGenerator) map[string]crypto.PrivateKey {
+	mapKeys := make(map[string]crypto.PrivateKey)
+
+	for i := uint16(0); i < 400; i++ {
```
Comment (sstanculeanu) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:45]: extract 400 into a constant

File: consensus/spos/bls/blsWorker_test.go (hunk 37 above)
Comment (sstanculeanu) [reply consensus/spos/bls/blsWorker_test.go:123]: this is not resolved, I was suggesting a new line before return

File: consensus/spos/bls/benchmark_verify_signatures_test.go (hunk 10 above)
Comment (sstanculeanu) [reply consensus/spos/bls/benchmark_verify_signatures_test.go:58]: not resolved

File: consensus/spos/bls/subroundEndRound.go (hunk 16 above)
Comment (sstanculeanu) [reply consensus/spos/bls/subroundEndRound.go:563]: unresolved

File: consensus/spos/bls/subroundEndRound.go [hunk 40]
```diff
@@ -553,7 +560,49 @@ func (sr *subroundEndRound) aggregateSigsAndHandleInvalidSigners(bitmap []byte)
...
+
+		sr.PeerHonestyHandler().ChangeScore(
+			pk,
+			spos.GetConsensusTopicID(sr.ShardCoordinator()),
+			decreaseFactor,
+		)
+		return err
+	}
+	log.Trace("verifyNodesOnAggSigVerificationFail: verifying signature share", "public key", pk)
+	return nil
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:600]:
```suggestion
	}
	
	log.Trace("verifyNodesOnAggSigVerificationFail: verifying signature share", "public key", pk)
	
	return nil
```

File: consensus/spos/bls/subroundEndRound_test.go [hunk 41]
```diff
@@ -361,12 +411,49 @@ func TestSubroundEndRound_NewSubroundEndRoundNilSyncTimerShouldFail(t *testing.T
...
+		sr,
+		bls.ProcessingThresholdPercent,
+		&statusHandler.AppStatusHandlerStub{},
+		&testscommon.SentSignatureTrackerStub{},
+		&mock.SposWorkerMock{},
+		nil,
+	)
+
+	assert.True(t, check.IfNil(srEndRound))
+	assert.NotNil(t, err)
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound_test.go:454]:
```suggestion
	assert.Equal(t, spos.ErrNilThrottler, err)
```

specific error should be checked

File: consensus/spos/bls/subroundEndRound_test.go [hunk 42]
```diff
@@ -1260,8 +1350,9 @@ func TestVerifyNodesOnAggSigVerificationFail(t *testing.T) {
 
 		sr.Header = &block.Header{}
 		_ = sr.SetJobDone(sr.ConsensusGroup()[0], bls.SrSignature, true)
-
-		_, err := sr.VerifyNodesOnAggSigFail()
+		ctx, cancel := context.WithCancel(context.TODO())
+		_, err := sr.VerifyNodesOnAggSigFail(ctx)
+		cancel()
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound_test.go:1355]:
```suggestion
		_, err := sr.VerifyNodesOnAggSigFail(context.TODO())
```
reason: test only, should be ok here.. valid for all tests below

File: consensus/spos/bls/subroundEndRound.go [hunk 43]
```diff
@@ -572,29 +623,29 @@ func (sr *subroundEndRound) verifyNodesOnAggSigFail() ([]string, error) {
...
-			invalidPubKeys = append(invalidPubKeys, pk)
-		}
+		wg.Add(1)
 
-		log.Trace("verifyNodesOnAggSigVerificationFail: verifying signature share", "public key", pk, "is successfull", isSuccessfull)
+		go func(i int, pk string, wg *sync.WaitGroup, sigShare []byte) {
+			defer func() {
+				defer sr.signatureThrottler.EndProcessing()
+				defer wg.Done()
+			}()
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound.go:639]:
```suggestion
			defer func() {
				sr.signatureThrottler.EndProcessing()
				wg.Done()
			}()
```

they are already in a defer func, no need to defer again

**Review by sstanculeanu — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

File: consensus/spos/bls/subroundEndRound_test.go [hunk 44]
```diff
@@ -62,6 +64,45 @@ func initSubroundEndRoundWithContainer(
 		appStatusHandler,
 		&testscommon.SentSignatureTrackerStub{},
 		&mock.SposWorkerMock{},
+		&dataRetrieverMocks.ThrottlerStub{},
+	)
+
+	return srEndRound
+}
+
+func initSubroundEndRoundWithContainer400Sig(
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound_test.go:73]: maybe rename the method - I see no constraint over there being 400 sigs.
I could call this method for a dfferent number of signatures, and it would be misleading then.
Could be renamed to `initSubroundEndRoundWithContainerAndConsensusState`

**Review by danielradu10 — APPROVED** [review]

File: consensus/spos/bls/subroundEndRound.go [hunk 45]
```diff
@@ -572,29 +623,29 @@ func (sr *subroundEndRound) verifyNodesOnAggSigFail() ([]string, error) {
...
-		err = sr.SigningHandler().VerifySignatureShare(uint16(i), sigShare, sr.GetData(), sr.Header.GetEpoch())
+		err = sr.checkGoRoutinesThrottler(ctx)
 		if err != nil {
-			isSuccessfull = false
-
-			err = sr.SetJobDone(pk, SrSignature, false)
-			if err != nil {
-				return nil, err
-			}
+			return nil, err
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:628]: on the returns here and at line 623 I think it is possible to panic inside the already running goroutines, as they will no longer have access to the context mutex and invalidPubKeys.

Could you create a test to validate/invalidate this?

File: consensus/spos/bls/subroundEndRound.go [hunk 46]
```diff
@@ -621,7 +672,9 @@ func (sr *subroundEndRound) getFullMessagesForInvalidSigners(invalidPubKeys []st
 }
 
 func (sr *subroundEndRound) handleInvalidSignersOnAggSigFail() ([]byte, []byte, error) {
-	invalidPubKeys, err := sr.verifyNodesOnAggSigFail()
+	ctx, cancel := context.WithTimeout(context.Background(), sr.RoundHandler().TimeDuration())
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/subroundEndRound.go:675]: in case of timeout, we need to make sure that the goroutines are not panicking, check my previous comment.

File: consensus/spos/bls/benchmark_verify_signatures_test.go [hunk 47]
```diff
@@ -0,0 +1,122 @@
...
+
+	signingHandler, err := factoryCrypto.NewSigningHandler(args)
+	require.Nil(b, err)
+
+	container.SetSigningHandler(signingHandler)
+	consensusState := initConsensusStateWithArgsVerifySignature(keysHandlerMock, keys)
+	dataToBeSigned := []byte("message")
+	consensusState.Data = dataToBeSigned
+
+	sr := *initSubroundEndRoundWithContainer400Sig(container, &statusHandler.AppStatusHandlerStub{}, consensusState, &dataRetrieverMocks.ThrottlerStub{})
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/benchmark_verify_signatures_test.go:109]: can you remove the * from the function call?

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: consensus/spos/bls/errors.go [hunk 48]
```diff
@@ -4,3 +4,6 @@ import "errors"
 
 // ErrNilSentSignatureTracker defines the error for setting a nil SentSignatureTracker
 var ErrNilSentSignatureTracker = errors.New("nil sent signature tracker")
+
+// ErrAux defines the error that does not have any meaning but helps for debug
```
Comment (AdoAdoAdo) [inline consensus/spos/bls/errors.go:8]: can you move this in the test file, this is only used in tests.

File: consensus/spos/bls/subroundEndRound_test.go [hunk 49]
```diff
@@ -1390,7 +1390,7 @@ func TestVerifyNodesOnAggSigVerificationFail(t *testing.T) {
 
 	t.Run("fail to verify signature share, an element will return an error on SignatureShare, should not panic", func(t *testing.T) {
 		t.Parallel()
-
+		var ErrAux = errors.New("auxiliary error")
```
Comment (sstanculeanu) [inline consensus/spos/bls/subroundEndRound_test.go:1393]: why not use expectedErr everywhere instead of this

**Review by sstanculeanu — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]
